src/app/clusters/tls-certificate-management-server/tls-certificate-management-server.cpp

```diff
@@ -0,0 +1,416 @@
+    if (result == CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+    }
+    else if (result != CHIP_NO_ERROR)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+    }
+}
+
+void TlsCertificateManagementServer::HandleRemoveRootCertificate(HandlerContext & ctx,
+                                                                 const RemoveRootCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: RemoveRootCertificate");
+    auto result = mDelegate.RemoveRootCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req.caid);
+    if (result == CHIP_ERROR_NOT_FOUND)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::NotFound);
+    }
+    else if (result == CHIP_NO_ERROR)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Success);
+    }
+    else
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+    }
+}
+
+void TlsCertificateManagementServer::HandleGenerateClientCsr(HandlerContext & ctx, const TLSClientCSR::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: TLSClientCSR");
+
+    VerifyOrReturn(req.nonce.size() <= 128, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+
+    auto status = mDelegate.GenerateClientCsr(
+        ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req, [&](auto & response) -> Status {
+            VerifyOrDieWithMsg(response.ccdid <= kMaxClientCertId, NotSpecified, "Spec requires CCDID to be <= kMaxClientCertId");
+            VerifyOrDieWithMsg(response.csr.size() <= 3000, NotSpecified, "Spec requires csr.size() to be <= 3000");
+            VerifyOrDieWithMsg(response.nonce.size() <= 128, NotSpecified, "Spec requires nonce.size() to be < kMaxClientCertId");
+            ctx.mCommandHandler.AddResponse(ctx.mRequestPath, response);
+            return Status::Success;
+        });
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+    }
+}
+
+void TlsCertificateManagementServer::HandleProvisionClientCertificate(HandlerContext & ctx,
+                                                                      const ProvisionClientCertificate::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "TlsCertificateManagement: ProvisionClientCertificate");
+
+    VerifyOrReturn(req.ccdid <= kMaxClientCertId, ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+    const auto & detail = req.clientCertificateDetails;
+    VerifyOrReturn(!detail.clientCertificate.HasValue() || detail.clientCertificate.Value().size() <= kSpecMaxCertBytes,
+                   ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError));
+
+    auto status = mDelegate.ProvisionClientCert(ctx.mRequestPath.mEndpointId, ctx.mCommandHandler.GetAccessingFabricIndex(), req);
```

Comment:
Same thing with client certificate provisioning, the existing cert should be checked before forwarding to the delegate and should return the spec defined error. 

In the current form, it ends up allocating duplicate certificate entries and allocating a new ID to each of them.

Reply:
That's not correct; `ProvisionClientCert` takes the ID from `TLSClientCSR`; if the ID has already been consumed by a previous `ProvisionClientCert`, then it acts as an update to an existing client cert, i.e. always calls `UpdateClientCertificateEntry`; this is per spec: "or rotate an existing client certificate."